### PR TITLE
Provide access to _type in 5.x indices

### PR DESCRIPTION
**Pull request by ywelsch**
Allows running queries against _type on 5.x indices as well as returning _type in search results.

Relates #81210

### CODE REVIEW

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/rest-compatibility (unrelated failure, as usual these days)

File: server/src/main/java/org/elasticsearch/index/mapper/LegacyTypeFieldMapper.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+    private static final LegacyTypeFieldMapper INSTANCE = new LegacyTypeFieldMapper();
+
+    public static final TypeParser PARSER = new FixedTypeParser(c -> INSTANCE);
+
+    protected LegacyTypeFieldMapper() {
+        super(new LegacyTypeFieldType(), Lucene.KEYWORD_ANALYZER);
+    }
+
+    public static class Defaults {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/LegacyTypeFieldMapper.java:40]: I don't think we need this block?

File: x-pack/qa/repository-old-versions/src/test/java/org/elasticsearch/oldrepos/OldRepositoryAccessIT.java [hunk 2]
```diff
@@ -442,6 +453,24 @@ private void assertDocs(String index, int numDocs, Set<String> expectedIds, Rest
...
+                );
+                logger.info(searchResponse);
+                assertEquals(typeCount, searchResponse.getHits().getTotalHits().value);
+                for (SearchHit hit : searchResponse.getHits().getHits()) {
+                    DocumentField typeField = hit.field("_type");
+                    assertNotNull(typeField);
+                    assertThat(typeField.getValue(), instanceOf(String.class));
+                    assertEquals(randomType, typeField.getValue());
+                }
+            }
```
Comment (romseygeek) [inline x-pack/qa/repository-old-versions/src/test/java/org/elasticsearch/oldrepos/OldRepositoryAccessIT.java:473]: 5x indices could contain multiple mapping types, do we have any tests that check that these can still be opened and searched?

**Review by romseygeek — APPROVED** [review]
LGTM, I left one question and one nitpick.

File: x-pack/qa/repository-old-versions/src/test/java/org/elasticsearch/oldrepos/OldRepositoryAccessIT.java (hunk 2 above)
Comment (ywelsch) [reply x-pack/qa/repository-old-versions/src/test/java/org/elasticsearch/oldrepos/OldRepositoryAccessIT.java:473]: This is exactly what this test is doing? When running against 5.x indices, multiple types are created, and we run searches against the index, filtering on a given type, as well as checking that the _type information is returned.

Comment (romseygeek) [reply x-pack/qa/repository-old-versions/src/test/java/org/elasticsearch/oldrepos/OldRepositoryAccessIT.java:473]: Ah right, sorry, I didn't get the context of the whole test.  LGTM then!

**Comment by ywelsch** [comment]
Thanks @romseygeek!

**Comment by jtibshirani** [comment]
@ywelsch could you give the background on why we're restoring support for searching and returning `_type`?

**Comment by ywelsch** [comment]
Sure, it's for accessing archival data (#81210) where users will have the ability to access their older raw data. Here, _type is part of the "original data" that the user provided to the system (but isn't stored in _source, similar to _id), so this PR adds support for them to be able to retrieve this data and to run basic queries on it.

**Comment by jtibshirani** [comment]
Got it, so we're adding some minimal handling for `_type` as this could store valuable data for 5.x indices. One other question -- I assume this re-adds a top-level `_type` key to the search response. Does this interact okay with 7.x compatibility, which also adds this (https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/SearchHit.java#L627-L629)?

**Comment by ywelsch** [comment]
> Got it, so we're adding some minimal handling for _type as this could store valuable data for 5.x indices. One other question -- I assume this re-adds a top-level _type key to the search response.

Correct. This is handled quite nicely in newer ES versions, which do that (top-level key) automatically for any metadata mapper.

> Does this interact okay with 7.x compatibility, which also adds this (https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/SearchHit.java#L627-L629)?

Interesting. This interacts badly, as the `_type` key would be added twice now (the existing RestApiVersion check should have accounted for that possibility). I can follow-up with a small fix here (low priority).

**Pull request merged**